### PR TITLE
Use migration tool to handle system changes

### LINKED ISSUE #726: Use DB migrations to handle changes to the DB during semester
Right now, whenever we touch the database, we have a somewhat disjointed process of applying the changes in that there is the potential to have the DB/site in an incompatible state (this happened to OS this semester at one point I believe). There's also no set way for developers to introduce changes to the DB (who handles writing the script to apply the changes? Who runs it? When?)

A way to handle this is to add a DB migration tool like [Flyway ](https://flywaydb.org/) or [Liquibase](http://www.liquibase.org/) with the tool running during the INSTALL_Submitty.sh script (write a hook that scans current semester course directory, uses them to check DBs) so that we don't need to worry about how we're going to make the changes on the DB with these changes (and it's easy to test on Vagrant).

Thoughts/suggestions @bmcutler @JerameyATyler?

### PR DESCRIPTION
**Pull request by MasterOdin**
This tool is to be used instead of dumping the .sql file into the master/course DBs at the start of the semester, and then adding onto `.setup/update_database.py` through the semester. This is messy and potentially error prone as it makes it hard to tell where the state of a given DB is at a given time. We should be using [database migrations](https://en.wikipedia.org/wiki/Schema_migration) to do this which is that we check in files that contains the database and then keep adding files for each new change we make. The tool then runs through all of these migration files, running the queries as necessary.

Due to the split nature of the database structure, it's hard to naturally implement an existing tool as we'd still have to write some amount of wrappers for the tool. For now, I've implemented a barebones migration tool that should handle this for us.

Usage is:
```
$ ./migration/migrator.py --help
usage: migrator.py [-h] [--environment {course,system,master}] --config
                   CONFIG_PATH
                   command ...

Migration script for upgrading/downgrading the database

positional arguments:
  command
    create              Create migration
    migrate             Run migrations
    rollback            Rollback the previously done migration

optional arguments:
  -h, --help            show this help message and exit
  --environment {course,system,master}, -e {course,system,master}
  --config CONFIG_PATH, -c CONFIG_PATH

$ ./migration/migrator.py create --help
usage: migrator.py create [-h] name

positional arguments:
  name        Name of argument

optional arguments:
  -h, --help  show this help message and exit

$ ./migration/migrator.py migrate --help
usage: migrator.py migrate [-h] [--fake] [--initial]

optional arguments:
  -h, --help  show this help message and exit
  --fake      Mark migrations as run without actually running them
  --initial   Only run the first migration and then mark the rest as done
              without running them.

$ ./migration/migrator.py rollback --help
usage: migrator.py rollback [-h]

optional arguments:
  -h, --help  show this help message and exit
```

migrator.py will run migrations out of the `migrations/` folder. However, if it detects migrations in the DB that it cannot match up to a file, it will attempt to clean out those migrations if it can find the file in `/usr/local/submitty/migrations` and then run the `down` function in that file before deleting the migration. `INSTALL_SUBMITTY_HELPER.sh` will rsync all files from `migration/migrations` to `/usr/local/submitty/migrations`. You can then use `create` to generate a template for a given environment (or all environments) that will have the filename `YYYYMMDDhhmmss_name.py` where the first bit should be a unique timestamp down to the second and the second part is a human readable string (given by the positional argument `name`). Status of whether or not a given migration file has been run is stored within a table in each database we touch (`migrations_{}` where {} is the environment name). Each file should have two functions `up` and `down` which takes one parameter (`connection`) which represents our DB connection for the migration if the environment is either master or course, else it takes no parameters. Each migration represents upgrading or downgrading a given environment.

Given environments:
* system - migrations for the submitty system at large (installing packages, etc.)
* master - migrations for the master submitty DB
* course - migrations that are run against each course DB

Closes #726.

### CODE REVIEW

**Comment by MasterOdin** [comment]
After discussion with @bmcutler, this PR is on hold till summer starts.

**Comment by bmcutler** [comment]
It looks like we've got 2 copies of the .sql files?

**Comment by MasterOdin** [comment]
@bmcutler, this should be ready for a new review from you/integration (after I'm done with my current `vagrant up` cycle). After your review and any requested changes are finalized, I'll write up the documentation for the developer site.

**Comment by MasterOdin** [comment]
@bmcutler, updated the README.md to fully cover the tool and its usage. Please let me know if you'd like anything else. I've also tested and fixed some bugs handling the case of missing migrations. It will now delete the file from `${SUBMITTY_INSTALL_DIR}/migrations` after it's `down` function has been run.

File: migration/README.md [hunk 1]
```diff
@@ -39,15 +44,60 @@ The tool has three commands:
...
 against the DB to get the status of the migrations. For any migration that is in
 the DB, but not in the folder, it will attempt to see if the migration exists
 in `${SUBMITTY_INSTALL_DIR}/migrations` and if so, will run the `down` function
-on each missing migration, and then delete the record of it from the DB.
+on each missing migration, and then delete the record of it from the DB as well as
+remove that migration file from `${SUBMITTY_INSTALL_DIR}/migrations`.
  
-`migrate` accepts two flags: `--fake` and `--initial`. Fake will mark all migrations
-as being run, without actually running them. `--initial` 
+`migrate` accepts two flags: `--fake` and `--initial`. `--fake` will mark all down migrations
```
Comment (bmcutler) [inline migration/README.md]: do you mean all *up* migrations will be marked as having been run?

File: migration/migrator.py [hunk 2]
```diff
@@ -202,9 +211,9 @@ def run_migration(connection, migration, environment, args):
 
 
 def load_migration_module(name, path):
-    spec = importlib.util.spec_from_file_location('module.{}'.format(name), str(path))
-    module = importlib.util.module_from_spec(spec)
-    spec.loader.exec_module(module)
+    # TODO: change this to not use depreciated loader.load_module after dropping Python 3.4 support
```
Comment (bmcutler) [inline migration/migrator.py]: deprecated not depreciated

Comment (MasterOdin) [reply migration/migrator.py]: I hate how similar these words are to each other and my inability for my brain to handle it.

**Comment by bmcutler** [comment]
Can you also revise the open line of the pull request message?

And can we add somewhere that we if starting from an even older install/database, you should first run install_system.sh  & update_database.py (for the last time!) before doing this migration stuff?

**Comment by bmcutler** [comment]
Oh, I think I see the migration table gets created if it doesn't exist, so no update_database step needed?  (right?)

**Comment by MasterOdin** [comment]
We might want to cut a new release that says "this is the final time you need to run update_database.py, after this, migrator.py will handle things for you"? Dunno where we might want to put that.

We will actually need to run `migrator.py migrate --fake` in update_database.py just to register the migrations as being "done" in the DB (as the initial migration creates the full schema).

**Comment by bmcutler** [comment]
Yes, please do that :)
Can we actually check & create the migrations table if doesn't exist in update_database?  Then add the migration table to the .sql file, then don't need to check & create if doesn't exist?  right?

**Comment by bmcutler** [comment]
And I was thinking that we should tag the migrations commit.  I'm thinking 1.2.0
Or should we move to year month numbering now and make it 18.06.00?
(I don't think we need the day in there.  If we every have multiple releases in a month we can use any numbering scheme).

**Pull request merged**